### DISCUSSION TITLE
span end and stopping/cleaning up the tracing

### DISCUSSION THREAD

**Discussion by kristianmo** [open-telemetry/opentelemetry-cpp#1215] [Q&A]
After the inital setup I get a tracer. Then in the http handler I create a span and a scope.

```c++
auto tracer = traceProvider->GetTracer("MyManager");
http::server rest_server;

rest_server.param<int>().get("/ping", [&tracer, &pingCounter](auto& req, auto& session, auto& args) {
		auto span = tracer->StartSpan("/ping", { { OTEL_GET_TRACE_ATTR(AttrHttpMethod), "GET" }, { OTEL_GET_TRACE_ATTR(AttrHttpScheme), "http" } });
		auto scope = tracer->WithActiveSpan(span);
		pingCounter.Increment();
		session.do_write(Http::SendHttpReply(req, int(boost::beast::http::status::ok), "true"));
});
```
The examples show a `span->End()` before we leave the scope of the span(as below). Is this needed in this pattern or does the created scope variable trigger a span cleanup(above)?

```c++
auto tracer = traceProvider->GetTracer("MyManager");
http::server rest_server;

rest_server.param<int>().get("/ping", [&tracer, &pingCounter](auto& req, auto& session, auto& args) {
		auto span = tracer->StartSpan("/ping", { { OTEL_GET_TRACE_ATTR(AttrHttpMethod), "GET" }, { OTEL_GET_TRACE_ATTR(AttrHttpScheme), "http" } });
		auto scope = tracer->WithActiveSpan(span);
		pingCounter.Increment();
		session.do_write(Http::SendHttpReply(req, int(boost::beast::http::status::ok), "true"));
                span->End();
});
```

Secondly, assuming this pattern in a thread:
```c++
//setup tracing
auto exporter = ...
auto resource = ...
auto processor = ...
auto sampler = ...
auto context = ...
auto traceProvider = ...
auto tracer = ...

// integrate with service calls
...

// wait for stop
server.WaitForStop()

// cleanup
exporter->Shutdown();
processor->Shutdown();
context->Shutdown();
tracer->Close(std::chrono::seconds(1));
```
Do we need to call the shutdown and close functions before exiting?

**Comment by lalitb** [comment]
> The examples show a span->End() before we leave the scope of the span(as below). Is this needed in this pattern or does the created scope variable trigger a span cleanup(above)?

span->End() is optional in the example posted by you, as it will implicitly called when the lambda function finishes ( through Span object's destructor). In general, it's good to call `span->End()` when there could be multiple references of Span object existing within the application, and we want to force the export of Span without waiting for all those references to get destroyed.

> Do we need to call the shutdown and close functions before exiting?

tracerProvider->Shutdown() is enough to shut down all other dependent objects ( exporter, processor and context ). Closing tracer is optional too.
